src/contrib/cluster/Akka.DistributedData/Replicator.cs

```diff
@@ -479,35 +486,91 @@ private void Load()
-            Receive<DeltaPropagationTick>(_ => ReceiveDeltaPropagationTick());
-            Receive<GossipTick>(_ => ReceiveGossipTick());
-            Receive<ClockTick>(c => ReceiveClockTick());
-            Receive<Internal.Status>(s => ReceiveStatus(s.Digests, s.Chunk, s.TotalChunks));
-            Receive<Gossip>(g => ReceiveGossip(g.UpdatedData, g.SendBack));
-            Receive<Subscribe>(s => ReceiveSubscribe(s.Key, s.Subscriber));
-            Receive<Unsubscribe>(u => ReceiveUnsubscribe(u.Key, u.Subscriber));
-            Receive<Terminated>(t => ReceiveTerminated(t.ActorRef));
-            Receive<ClusterEvent.MemberWeaklyUp>(m => ReceiveMemberWeaklyUp(m.Member));
-            Receive<ClusterEvent.MemberUp>(m => ReceiveMemberUp(m.Member));
-            Receive<ClusterEvent.MemberRemoved>(m => ReceiveMemberRemoved(m.Member));
-            Receive<ClusterEvent.IMemberEvent>(m => ReceiveOtherMemberEvent(m.Member));
-            Receive<ClusterEvent.UnreachableMember>(u => ReceiveUnreachable(u.Member));
-            Receive<ClusterEvent.ReachableMember>(r => ReceiveReachable(r.Member));
-            Receive<GetKeyIds>(_ => ReceiveGetKeyIds());
-            Receive<Delete>(d => ReceiveDelete(d.Key, d.Consistency, d.Request));
-            Receive<RemovedNodePruningTick>(r => ReceiveRemovedNodePruningTick());
-            Receive<GetReplicaCount>(_ => ReceiveGetReplicaCount());
+            switch (message)
+            {
+                case IDestinationSystemUid msg:
+                    if (msg.ToSystemUid.HasValue && msg.ToSystemUid != _selfUniqueAddress.Uid)
+                    {
+                        // When restarting a node with same host:port it is possible that a Replicator on another node
+                        // is sending messages to the restarted node even if it hasn't joined the same cluster.
+                        // Therefore we check that the message was intended for this incarnation and otherwise
+                        // it is discarded.
+                        _log.Info("Ignoring message [{0}] from [{1}] intended for system uid [{2}], self uid is [{3}]",
+                            Logging.SimpleName(msg),
+                            Sender,
+                            msg.ToSystemUid,
+                            _selfUniqueAddress.Uid);
+                    }
+                    else
+                    {
+                        switch (msg)
+                        {
+                            case Status s:
+                                ReceiveStatus(s.Digests, s.Chunk, s.TotalChunks); return true;
+                            case Gossip g : 
+                                ReceiveGossip(g.UpdatedData, g.SendBack); return true;
+                        }
+                    }
+                    return true;
+
+                case ISendingSystemUid msg:
+                    if (msg.FromNode != null && !IsKnownNode(msg.FromNode.Address))
```

Comment:
Special case for `Read`, `Write`, and `DeltaPropagation`, we filter out messages from all unknown nodes. Reason is the same as above.